src/tests/Microsoft.PowerFx.Interpreter.Tests/MarshalTests.cs

```diff
@@ -0,0 +1,569 @@
+        private class PocoNotMarshalled
+        {
+            private int Prop1 { get; set; } // not public
+
+            protected int Prop2 { get; set; } // not public
+
+            public int _field; // not a property 
+
+            public static int PropStatic { get; set; } // not static
+
+            public int PropNoGet { set => _field = value; } // write-only 
+        }
+
+        [Fact]
+        public void NotMarshalled()
+        {
+            var cache = new TypeMarshallerCache();
+            var tm = cache.GetMarshaller(typeof(PocoNotMarshalled));
+            var fxType = (RecordType)tm.Type;
+
+            Assert.Empty(fxType.GetNames());
+        }
+
+        private interface IWidget
+        {
+            public string Data { get; }
+        }
+
+        public class Widget : IWidget
+        {
+            public string Data { get; set; }
+        }
+
+        // Custom marshaller. Marshal Widget objects as Strings with a "W" prefix. 
+        private class WidgetMarshalerProvider : ITypeMashallerProvider
+        {
+            public int _counter = 0;
+
+            public bool TryGetMarshaller(Type type, TypeMarshallerCache cache, int maxDepth, out ITypeMarshaller marshaler)
+            {
+                if (type != typeof(IWidget))
+                {
+                    marshaler = null;
+                    return false;
+                }
+
+                _counter++;
+
+                marshaler = new WidgetMarshaler();
+                return true;
+            }
+
+            private class WidgetMarshaler : ITypeMarshaller
+            {
+                public FormulaType Type => FormulaType.String;
+
+                public FormulaValue Marshal(object value)
+                {
+                    // adding a "W" prefix ensures this code is run and it's not just some cast. 
+                    var w = (Widget)value;
+                    return FormulaValue.New("W" + w.Data);
+                }
+            }
+        }
+
+        // Fails to marshal without the customer marshaller. 
+        [Fact]
+        public void FailWithoutCustomMarshaller()
+        {
+            var cache = new TypeMarshallerCache();
+
+            var obj = new
+            {
+                Length = 12.0,
+                Widget1 = (IWidget)new Widget
+                {
+                    Data = "A"
+                },
+                Widget2 = (IWidget)new Widget
+                {
+                    Data = "B"
+                }
+            };
+            Assert.Throws<InvalidOperationException>(() => cache.Marshal(obj));
+        }
+
+        // Test a custom marshaler. 
+        [Fact]
+        public void CustomMarshalerType()
+        {
+            var cache = new TypeMarshallerCache();
+
+            // Insert at 0 to get precedence over generic object marshaller
+            var marshaler = new WidgetMarshalerProvider();
+            cache.Marshallers.Insert(0, marshaler);
+
```

Comment:
This seems like a method we should provide. Exposing the list and requiring hosts to know about the order and the meaning of the order isn't ideal.